### PR TITLE
[7] Create API endpoint for GET Surveys 

### PR DESCRIPTION
**Pull request by junan**
https://github.com/nimblehq/liveman/projects/1#card-63294161

## What happened

Implemented `GET` Survey list API(`/v1/surveys`)
 
## Insight

- Implemented the response structure same as our survey list API
- Backend implementation also exactly the same as our survey list API(Parsed `surveys.json`) data, assigning them to embedded schema models(no saving in the database), the expose them
- Used [scrivener_list](https://github.com/stephenmoloney/scrivener_list) for pagination
## Proof Of Work

Screen Shot 2021-07-02 at 10 23 30 AM

Full response https://gist.github.com/junan/faa1e3e62464098f38b14232dc46e256

### CODE REVIEW

File: lib/liveman/survey/surveys.ex [hunk 1]
```diff
@@ -0,0 +1,70 @@
+defmodule Liveman.Survey.Surveys do
+  use Ecto.Schema
+
+  alias Liveman.Survey.Schemas.Question
+  alias Liveman.Survey.Schemas.Survey
+
+  @json_file "priv/repo/data/surveys.json"
+
+  def list_surveys do
+    survey_list_json = get_json()
```
Comment (bterone) [inline lib/liveman/survey/surveys.ex:10]: I don't really get how `get_json()` is returning a tuple `{:ok, json}` but it still can go through the `Enum.map` without pattern matching it here. 🤔 
Is there something I'm missing here? 😂

File: lib/liveman/survey/surveys.ex [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    Enum.map(survey_list_json, fn survey_json ->
+      survey = build_survey(survey_json)
+      questions = build_question(survey_json["questions"], survey)
+
+      survey
+      |> Ecto.Changeset.change()
+      |> Ecto.Changeset.put_embed(:questions, questions)
+      |> Ecto.Changeset.apply_changes()
```
Comment (bterone) [inline lib/liveman/survey/surveys.ex:19]: I wonder if we can use `cast_embed` instead. I think it would put the embed and apply the changes at the same time 🤔

File: lib/liveman/survey/surveys.ex (hunk 1 above)
Comment (junan) [reply lib/liveman/survey/surveys.ex:10]: Checked more, though it was working functionality, but I found that `Jason.decode!(body)` does not return a tuple, instead it returns the result directly and it cant match with ` {:ok, json}` and `with` case it returns the result if it cant match. Replaced `with` case with simple parsing code.

File: lib/liveman/survey/surveys.ex (hunk 2 above)
Comment (junan) [reply lib/liveman/survey/surveys.ex]: Checked more. In this feature, we actually directly build the survey embedded_schema including associated embedded_schema (`questions`) without `cast_embed` and then pass it to the JSON serializer. It works.

I still added `cast_embed` in case we need to update the survey schema with the association, it would be useful at that time.

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
 
 on: pull_request
 
 env:
-  OTP_VERSION: "24.0.1"
-  ELIXIR_VERSION: "1.12.1"
-  NODE_VERSION: "14"
+  OTP_VERSION: '24.0.1'
+  ELIXIR_VERSION: '1.12.1'
+  NODE_VERSION: '14'
```
Comment (olivierobert) [inline .github/workflows/test.yml:8]: Why the change though?

File: lib/liveman/survey/surveys.ex [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+  def list_surveys do
+    survey_list_json = get_json()
+
+    Enum.map(survey_list_json, fn survey_json ->
+      build_survey(survey_json)
+    end)
+  end
+
+  def get_json do
```
Comment (olivierobert) [inline lib/liveman/survey/surveys.ex:17]: In terms of naming, this method is not very descriptive. How about instead:

```suggestion
  def fecth_surveys_from_file do
```

File: lib/liveman_web/router.ex [hunk 5]
```diff
@@ -2,28 +2,29 @@ defmodule LivemanWeb.Router do
   use LivemanWeb, :router
 
   pipeline :browser do
-    plug :accepts, ["html"]
-    plug :fetch_session
-    plug :fetch_flash
-    plug :protect_from_forgery
-    plug :put_secure_browser_headers
+    plug(:accepts, ["html"])
```
Comment (olivierobert) [inline lib/liveman_web/router.ex:5]: Where are these formatting changes come from? 🤔 `mix.format` does not seem to require this so I am surprised.

File: mix.exs [hunk 6]
```diff
@@ -58,6 +58,7 @@ defmodule Liveman.MixProject do
       {:phoenix_live_reload, "~> 1.2", only: :dev},
       {:phoenix_live_dashboard, "~> 0.4"},
       {:plug_cowboy, "~> 2.0"},
+      {:scrivener_list, "~> 2.0"},
```
Comment (olivierobert) [inline mix.exs:61]: Why picking this library over https://github.com/elixirdrops/kerosene as it was suggested by Micky in this [Slack thread](https://nimble-co.slack.com/archives/C0112R61J3W/p1623640061002400?thread_ts=1623639276.000700&cid=C0112R61J3W)

Comment (junan) [reply mix.exs:61]: The `kerosene` library is for `Ecto`. But here we are not storing any data into the database(we are just fetching surveys data from the file, assign them to the `embedded_schema` and expose the list with `jsonapi`  just like how our rails survey app does). 

We get here only survey `list` and the `scrivener_list` library is just for `list` pagination. like bellow

```elixir
[ "Elixir", "Ruby", "Go", "JavaScript", "PHP",  "Perl", "Python"] |> Scrivener.paginate(%{})
```

Whst do you think?

File: .github/workflows/test.yml (hunk 3 above)
Comment (junan) [reply .github/workflows/test.yml]: Reverted, my VS code `Prettier` plugin did it on save 😄

File: lib/liveman_web/router.ex (hunk 5 above)
Comment (junan) [reply lib/liveman_web/router.ex]: Reverted, I think my VS code `ElixirLS` plugin did it on save 😄

File: lib/liveman/survey/surveys.ex [hunk 7]
```diff
@@ -0,0 +1,66 @@
+defmodule Liveman.Survey.Surveys do
+  use Ecto.Schema
+
+  alias Liveman.Survey.Schemas.Question
+  alias Liveman.Survey.Schemas.Survey
```
Comment (rosle) [inline lib/liveman/survey/surveys.ex:5]:
```suggestion
  alias Liveman.Survey.Schemas.{Question, Survey}
```

File: test/liveman/survey/schemas/survey_test.exs [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  describe "changeset/2" do
+    test "returns valid changeset if given valid params" do
+      attr = params_for(:survey)
+
+      changeset = Survey.changeset(%Survey{}, attr)
+
+      assert changeset.valid?
+    end
+
+    test "returns invalid changeset if params are missing" do
```
Comment (rosle) [inline test/liveman/survey/schemas/survey_test.exs:15]:
```suggestion
    test "returns invalid changeset if required params are missing" do
```

File: test/liveman/user/schemas/question_test.exs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+  describe "changeset/2" do
+    test "returns valid changeset if given valid params" do
+      attr = params_for(:question)
+
+      changeset = Question.changeset(%Question{}, attr)
+
+      assert changeset.valid?
+    end
+
+    test "returns invalid changeset if params are missing" do
```
Comment (rosle) [inline test/liveman/user/schemas/question_test.exs:15]:
```suggestion
    test "returns invalid changeset if required params are missing" do
```

File: test/liveman_web/requests/v1/survey_request_test.exs [hunk 10]
```diff
@@ -0,0 +1,17 @@
+defmodule LivemanWeb.V1SurveyRequestTest do
```
Comment (rosle) [inline test/liveman_web/requests/v1/survey_request_test.exs:1]:
```suggestion
defmodule LivemanWeb.V1.SurveyRequestTest do
```

File: test/liveman_web/requests/v1/survey_request_test.exs [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  use LivemanWeb.ConnCase, async: true
+
+  describe "GET /v1/surveys" do
+    test "returns 200 status when request successfully", %{conn: conn} do
+      conn = get(conn, Routes.api_v1_survey_path(conn, :index))
+
+      assert conn.status == 200
+    end
+
+    test "returns 200 status when request with page_number and  page_size", %{conn: conn} do
```
Comment (rosle) [inline test/liveman_web/requests/v1/survey_request_test.exs:11]:
```suggestion
    test "returns 200 status when request with page_number and page_size", %{conn: conn} do
```

File: mix.exs (hunk 6 above)
Comment (olivierobert) [reply mix.exs:61]: That looks fine 😁  I was just wondering about it 👍

File: lib/liveman/survey/schemas/survey.ex [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  alias Liveman.Survey.Schemas.Question
+
+  @primary_key {:id, :string, autogenerate: false}
+  embedded_schema do
+    field(:access_code_prompt, :string)
+    field(:access_code_validation, :string)
+    field(:active_at, :naive_datetime)
+    field(:cover_background_color, :string)
+    field(:cover_image_url, :string)
+    field(:created_at, :naive_datetime)
```
[the remark this answers is not in the document]
Comment (junan) [reply lib/liveman/survey/schemas/survey.ex]: Used `timestamps(inserted_at: :created_at)`

File: lib/liveman/survey/surveys.ex [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+      thank_email_above_threshold: json["thank_email_above_threshold"],
+      thank_email_below_threshold: json["thank_email_below_threshold"],
+      is_active: json["is_active"],
+      cover_image_url: json["cover_image_url"],
+      created_at: json["created_at"],
+      active_at: json["active_at"],
+      inactive_at: json["inactive_at"],
+      type: json["type"],
+      questions: build_question(json["questions"], json["id"])
+    }
```
[the remark this answers is not in the document]
Comment (junan) [reply lib/liveman/survey/surveys.ex]: It works! thanks 👍

File: lib/liveman/survey/surveys.ex [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+  def list_surveys do
+    survey_list_json = fecth_surveys_from_file!()
+
+    Enum.map(survey_list_json, fn survey_json ->
+      build_survey(survey_json)
+    end)
+  end
+
+  defp fecth_surveys_from_file! do
```
Comment (rosle) [inline lib/liveman/survey/surveys.ex:16]:
```suggestion
  defp fetch_surveys_from_file! do
```

File: test/support/factory.ex [hunk 15]
```diff
@@ -2,4 +2,6 @@ defmodule Liveman.Factory do
   use ExMachina.Ecto, repo: Liveman.Repo
 
   use Liveman.User.UserFactory
+  # credo:disable-for-next-line Credo.Check.Consistency.MultiAliasImportRequireUse
```
Comment (rosle) [inline test/support/factory.ex:5]: I think we can disable this, 🤔  Maybe we should raise in the Elixir template cause we mostly do multi alias. Not sure maybe we can recheck with other Elixir projects 🤔  - https://github.com/rrrene/credo/blob/master/.credo.exs#L169

**Review by rosle — APPROVED** [review]

**Review by bterone — APPROVED** [review]

File: test/support/factory.ex (hunk 15 above)
Comment (junan) [reply test/support/factory.ex]: I disabled the config for now. It mostly happens initial changes of the project but later when we use more multi alias, then `credo` does not show the warning. The Elixir projects(mostly internal) I can access, we did not disable the config nor comment in the line.

I will check it more and will raise issue in the template
